nixjs-rt/src/errors/index.ts

```diff
@@ -0,0 +1,196 @@
+import {
+  Attrset,
+  Lambda,
+  Lazy,
+  NixBool,
+  NixFloat,
+  NixInt,
+  NixList,
+  NixNull,
+  NixString,
+  NixType,
+  NixTypeClass,
+  NixTypeInstance,
+  Path,
+} from "../lib";
+import { NixAbortError } from "./abort";
+import {
+  NixAttributeAlreadyDefinedError,
+  NixMissingAttributeError,
+} from "./attribute";
+import { NixFunctionCallWithoutArgumentError } from "./function";
+import { NixOtherError } from "./other";
+import { NixTypeMismatchError } from "./typeError";
+import { NixCouldntFindVariableError } from "./variable";
+
+/** A helper to convert instances to their respective class in a type-safe way, for better error messages */
+export function instanceToClass(instance: NixTypeInstance | NixTypeClass) {
+  if (instance instanceof NixBool) {
+    return NixBool;
+  } else if (instance instanceof NixFloat) {
+    return NixFloat;
+  } else if (instance instanceof NixInt) {
+    return NixInt;
+  } else if (instance instanceof NixList) {
+    return NixList;
+  } else if (instance instanceof NixNull) {
+    return NixNull;
+  } else if (instance instanceof NixString) {
+    return NixString;
+  } else if (instance instanceof Path) {
+    return Path;
+  } else if (instance instanceof Lambda) {
+    return Lambda;
+  } else if (instance instanceof Attrset) {
+    return Attrset;
+  } else if (instance instanceof Lazy) {
+    return instanceToClass(instance.toStrict());
+  } else {
+    return instance;
+  }
+}
+
+type PlainErrorMessagePart = {
+  kind: "plain";
+  value: string;
+};
+
+type HighlightedErrorMessagePart = {
+  kind: "highlighted";
+  value: string;
+};
+
+type ErrorMessagePart = PlainErrorMessagePart | HighlightedErrorMessagePart;
+export type ErrorMessage = ErrorMessagePart[];
+
+/** A hack-y way of finding whether an object is an ErrorMessagePart. Essential for error message building. */
+function isErrorMessagePart(part: any): part is ErrorMessagePart {
+  return (
+    typeof part === "object" &&
+    part !== null &&
+    "kind" in part &&
+    typeof part.kind === "string"
+  );
+}
+
+function isErrorMessage(parts: any): parts is ErrorMessage {
+  if (!Array.isArray(parts)) {
+    return false;
+  }
+
+  return parts.every(isErrorMessagePart);
+}
+
+type ErrorMessageBuilderPart =
+  | string
+  | NixTypeClass
+  | NixTypeClass[]
+  | NixTypeInstance
+  | ErrorMessage;
```

Comment:
A very nice idea to let us build error messages!

I've been staring at this a bit and I believe that it contains some assumptions that make it a bit harder to grasp. For example: the assumption that a nix type, list of nix types (which is concatenated with "or", rather than "and"), and a single nix type instance (but not nix type instance list) should be highlighted automatically. These assumptions also make the logic a bit rigid, complex, and hard to wrap one's head around.

How about we remove the assumptions, and simplify this API a bit to make it more composable and specializable?

Say we leave it up to the user to decide what is highlighted? It would then look like this:

```typescript
err`Expected ${boldAnyOf([NixInt, NixFloat])}, but got ${bold(value)}.`
```

This will give you an API with simpler logic that you can build on top of. I would say this is a reasonable compromise between simplicity, implicit behaviour, and UX. I hope that sounds reasonable.

Reply:
Should be added, lmn what you think